feat: word diffs

KnorpelSenf opened the pull request:
Adds a word diffing feature for each hunk. The following questions need to be answered to move this out of draft state:

> **EDIT**: all done!

1. The current implementation is based on bytes. This is consistent with how strings are used internally in other places. Is this useful for word-diffs? We could also use `.chars()` or `.graphemes(true)` (via [this crate](https://crates.io/crates/unicode-segmentation)). It's also possible to offer several of these options. (**EDIT**: we now use words)
2. The current implementation reuses the diff, but it reallocates the bytes of the underlying strings for every hunk. Is this acceptable? If no, which alternative should we go with? (**EDIT**: no longer applies since we no longer use bytes)
3. The current implementation reuses the diff, but it reallocates the `InternedInput` for every hunk. Is this acceptable? If no, which alternative should we go with? (**EDIT**: we now reuse the interner)
4. The current implementation estimates the token count to be 256 for byte token sources. Please let me know if that is an unreasonable heuristic for byte token sources. (**EDIT**: no longer applies since we no longer use bytes)
5. The current implementation always uses the Myers algo. Should we offer a second method which performs a minimal diff? (**EDIT**: no, we can add that later if requested)
6. Is there anything else I am missing?

Closes #1.

KnorpelSenf committed:
feat: define bytes token source

KnorpelSenf committed:
feat: implement byte-based word diff

KnorpelSenf committed:
test: cover word diffs

KnorpelSenf committed:
docs: fix typo

KnorpelSenf commented on src/lib.rs:
```diff
@@ -389,6 +392,35 @@ impl Hunk {
+    /// Performs a byte-diff of the hunk
+    pub fn byte_diff(&self, diff: &mut Diff, input: &InternedInput<&str>) {
+        let Hunk { before, after } = self.clone();
+        let before_bytes: &[u8] = &before
+            .flat_map(|token| input.interner[input.before[token as usize]].bytes())
+            .collect::<Vec<_>>();
+        let after_bytes: &[u8] = &after
+            .flat_map(|token| input.interner[input.after[token as usize]].bytes())
+            .collect::<Vec<_>>();
```
Points 1-2

pascalkuthe replied:
I definitly wouldn't diff bytes, I would probably diff words. Git diffs words and so does delta. Delta will (by default) essentially treat a run of word characters as a single word, Any other character is a word on its own

KnorpelSenf replied:
There are a million ways to tokenize a string into words. It is probably easy for ASCII, but for the rest of Unicode, I could not tell you how to do this properly. How should this be done in imara?

Byron replied:
It's interesting how we are going from needing ASCII newlines to tokenize to needing to understand the encoding of the byte buffer in full. `char::is_word()` doesn't seem to exist in `std`, but there [is a crate](https://docs.rs/unicode-segmentation/latest/unicode_segmentation/struct.UnicodeWords.html) that can certainly do it, at least for anything that can be decoded to unicode. To do that, one can use `chardetng` for encoding detection, and then one is just one `decode()` call away from trying to decode the buffer.

KnorpelSenf replied:
Yep, I got stuck yesterday when I realised that `char::is_word` isn't actually a thing, which brings me back to my original question. Many ways to split a string into words, which method to use? 

We can do bytes, we can do `chars`, we can use that crate (and I'm sure this isn't the only crate for the job), we can let the library consumer specify an implementation, we can offer several options and default to one, we can use a regex (what's the default of git's `--word-diff-regex` anyway), we could use grapheme clusters, we could simply ignore that there are grapheme clusters out there and build our own thing that focuses on ASCII, and I'm sure many of those ideas can be combined in several ways, too.

pascalkuthe replied:
Delta uses regex for word characters class but I wouldn't want to depend on regex for that. I missremebered about the character class. Rust does provide something that is almost good enough for a word: `char::is_alphanumeric` and then in helix we use this definiton: https://github.com/helix-editor/helix/blob/bfcbef10c513108c7b43317569416c2eefc4ed44/helix-core/src/chars.rs#L85

That works in well in practice and ist almost the same as regex (there may be some niche join control character classes). 

Probably for the rest graphemes are more appropriate unit of diff but I don't really want to introduce the dependency on unixode segmentation (or if we do do it we put this feature behind a feature flag). For many cases chars are good enough. Bytes are certainly not a good unit because that would mean the diff can produce non-utf8 aligned slices which would mean you cannot create a string slice with it.

Regarding encoding: Rust strings are always utf8 and I was assuming that this feature would work on those (so if you had non-utf8 strings you would use chardetcy and encodings-rs to reencode to utf8). I think if your text is an arbitrary encoding you need that anyway because other encoding are not necessarily a superset of asccii and segmentation into words is important both for good performance and good results

KnorpelSenf replied:
That sounds reasonable, thanks!

KnorpelSenf commented on src/lib.rs:
```diff
@@ -389,6 +392,35 @@ impl Hunk {
+        } else if self.is_pure_insertion() {
+            diff.added.fill(true);
+        } else {
+            let input = InternedInput::new(bytes(before_bytes), bytes(after_bytes));
```
Point 3

pascalkuthe replied:
We should reuse the interner

KnorpelSenf commented on src/sources.rs:
```diff
@@ -79,6 +84,37 @@ impl<'a> TokenSource for Lines<'a> {
+    }
+
+    fn estimate_tokens(&self) -> u32 {
+        0xFF
```
Point 4

KnorpelSenf commented on src/lib.rs:
```diff
@@ -389,6 +392,35 @@ impl Hunk {
+        } else {
+            let input = InternedInput::new(bytes(before_bytes), bytes(after_bytes));
+            diff.compute_with(
+                Algorithm::Myers,
```
Point 5

KnorpelSenf replied:
I am specifically referring to https://github.com/pascalkuthe/imara-diff/blob/8de65af4621b4d83c35b458a73f544cb68842edb/src/lib.rs#L221

pascalkuthe replied:
I suspect that word diffs are rarely large enough for that heuristic to kick in and if it does then you probably don't care about the results anyway (way bigger than a human could read). In general the minimal node is more aimed at usecases where you are not producing a readable diff but rather something for storage/network transfers (possibly not even riffing text). Word diffs are only really useful for text based diffs and probably usually human readable ones so I suspect it wouldn't be commonly used. 

With these kind of things I tend to take a eait and see approach: I would hardcode the algorithm until a real world usecase for other algorithms actually materializes

KnorpelSenf committed:
feat: tokenize words not bytes

KnorpelSenf committed:
test: cover word diffs

KnorpelSenf committed:
Merge branch 'master' into word-diff

KnorpelSenf committed:
docs: fix oversight

KnorpelSenf commented:
@pascalkuthe @Byron I have addressed all comments. This is ready for review now.

KnorpelSenf committed:
fix: lint about needless `&`

KnorpelSenf committed:
perf: no need to copy word vec

KnorpelSenf committed:
refactor: only branch on word length

KnorpelSenf committed:
perf: merge contiguous blocks of space chars

Byron commented on src/sources.rs:
```diff
@@ -79,6 +86,53 @@ impl<'a> TokenSource for Lines<'a> {
+                .char_indices()
+                .find(|(_, c)| *c != ' ')
+                .map_or(self.0.len(), |(index, _)| index)
+        } else if initial.is_alphanumeric() {
```
Independently of this (I lack competence here), I wonder if we *must* force this into `&str`?
It would be more usable/compatible if we'd assume only ASCII in `&[u8]`.

This should be fine, as most revision control systems either assume ASCII bytes themselves, or keep information about the encoding to allow decoding it into UTF-8 which is ASCII compatible.

KnorpelSenf replied:
> I wonder if we _must_ force this into `&str`?

What are you referring to by _this_?

I don't think my changes perform any conversion from `&[u8]` to `&str`, do they?

Byron replied:
This was meant to say that my comment isn't a reply to your comment 😅.

> I don't think my changes perform any conversion from `&[u8]` to `&str`, do they?

The idea is to have functions take `&[8]` so they work with unknown, but ascii-compatible encodings, and String inputs alike.
But I think my mistake is to remember the implementation of UnifiedDiffing in `gix-diff` which operates on bytes, whereas the one here may always have operated on `&str`, and that it wouldn't even be a good idea to change that in the same PR.
So I think I have to retract my comment, and apologise for the confusion.

KnorpelSenf replied:
Ahhh now I understood. Alright

KnorpelSenf committed:
docs: improve `words` docs w.r.t. space handling

KnorpelSenf committed:
test: cover words tokenizer

KnorpelSenf commented:
@pascalkuthe is there a way for to me make this easier to review?

KnorpelSenf committed:
fix: consider `_` as part of a word

Co-authored-by: Pascal Kuthe <pascalkuthe@pm.me>

KnorpelSenf commented:
Thanks! Can you allow CI to run?

KnorpelSenf commented on src/sources.rs:
```diff
@@ -79,6 +87,53 @@ impl<'a> TokenSource for Lines<'a> {
+        } else if initial.is_alphanumeric() {
+            self.0
+                .char_indices()
+                .find(|(_, c)| !c.is_alphanumeric())
```
I think I just misunderstood then.

> Also probably something similiar needed for the start chart

What did you mean by this?

KnorpelSenf commented on src/sources.rs:
```diff
@@ -79,6 +87,53 @@ impl<'a> TokenSource for Lines<'a> {
+        } else if initial.is_alphanumeric() {
+            self.0
+                .char_indices()
+                .find(|(_, c)| !c.is_alphanumeric() && c != '_')
```
```suggestion
                .find(|(_, c)| !c.is_alphanumeric() && *c != '_')
```
should fix it

KnorpelSenf committed:
fix: add missing deref

KnorpelSenf committed:
test: cover _ tokenization

KnorpelSenf committed:
test: remove lost test case

pascalkuthe commented:
ah sorry I misunderstood weird for me it shows 
<img width="843" height="373" alt="image" src="https://github.com/user-attachments/assets/79e7cf2d-1f32-4a32-8bd6-d144b4669e4c" />

KnorpelSenf commented:
Fascinating: 
![Screenshot_20251219-214900](https://github.com/user-attachments/assets/e2159dde-465e-4fc8-8b33-8e21ac6935fa)

Byron reviewed (changes requested):
Copilot did inspire me to look more closely, and I think some more work is needed.

* Let's make clear that we only deal with latin words here with a change to the function name.
* let's document the function and all of its parameters in detail. The `&mut` ones need extra work to make clear what the user can expect or is expected to do. Particularly what happens to the memory there, will it build up?

I'd go as far as to say that we need a fuzz-test for this one, so could block it on #42 or change the target to merge into that one. All these index operations I wouldn't trust until a fuzzer ran it with a couple million inputs.

Thanks again for making this happen @KnorpelSenf, I hope it's not too much of a burden to go into another round.

KnorpelSenf committed:
Merge branch 'master' into word-diff

KnorpelSenf committed:
docs: elaborate on `word_diff` usage

Byron committed:
refactor

- rename `word_diff` to `latin_word_diff`

Byron committed:
refactor tests

Move all tests that don't validate private state are on integration level.
Reorganise fixtures while at it.

Byron committed:
Add more tests for the word diff to understand postprocessing

Byron reviewed (approved):
Thanks a lot!

I made some more adjustments, mostly to the testing so I get it, too 😁.
